### ISSUE TITLE
Allow accessing other members of a class which implements IPropertyBag

### ISSUE THREAD

**Issue by KurtGokhan** [microsoft/ClearScript#269]
For example, following test fails. But if I remove `IPropertyBag`, it works as expected.

```csharp
        public class TestDictionary : Dictionary<string, object>, IPropertyBag
        {
            public string TestValue => "test1";
            public string TestMethod() => "test2";
        }

        [TestMethod]
        public void TestPropertyBagExtraProps()
        {
            engine.AddHostObject("dc", new TestDictionary());
            Assert.AreEqual("test1", engine.Evaluate("dc.TestValue"));
            Assert.AreEqual("test2", engine.Evaluate("dc.TestMethod()"));
        }
```

This is written in the documentation of `IPropertyBag`:

> If an object that implements this interface is added to a script engine, script code will be able to access the properties stored in the collection as if they were members of the object itself, using the script language's native syntax for member access. No other members of the object will be accessible. 

I want to know if this is by design, or a limitation by .Net? Is it possible to allow extra properties to be accessed if the property isn't found in the `PropertyBag`?

**Comment by ClearScriptLib** [comment]
Hi @KurtGokhan,

>I want to know if this is by design, or a limitation by .Net?

It's by design. The intention was to avoid ambiguity. Suppose `TestValue` is a writable property, and the script engine issues `dc.TestValue = 123`. Should this be an error, since `TestValue` requires a string argument? Or should it add a second `TestValue` property that hides the original? We prefer not to make arbitrary decisions on behalf of the application.

Instead, you can implement a strategy that works for you. Here's one possibility that leverages V8's capabilities:

``` C#
public class TestDictionaryBase : Dictionary<string, object> {
    public string TestValue => "test1";
    public string TestMethod() => "test2";
}
public class TestDictionary : TestDictionaryBase, IPropertyBag {}
```

And then:

``` C#
var dc = new TestDictionary();
engine.AddHostObject("dc", dc);
engine.Script.Object.setPrototypeOf(dc, dc.ToRestrictedHostObject<TestDictionaryBase>(engine));
```

With this setup, you can do things like this:

``` C#
Console.WriteLine(engine.Evaluate("dc.TestValue"));     // "test1"
Console.WriteLine(engine.Evaluate("dc.TestMethod()"));  // "test2"
engine.Execute("dc.TestValue = 123");
Console.WriteLine(engine.Evaluate("dc.TestValue"));     // 123
engine.Execute("delete dc.TestValue");
Console.WriteLine(engine.Evaluate("dc.TestValue"));     // "test1"
```

This is just an example. JavaScript gives you many ways to set up objects with custom behavior.

Good luck!

**Comment by KurtGokhan** [comment]
Woah, I didn't know you could use prototype like that with marshalled objects. Makes so much sense. 

My initial expectation was not met but seems like there is a pretty solid workaround.

**Comment by ClearScriptLib** [comment]
Hi @KurtGokhan,

>Woah, I didn't know you could use prototype like that with marshalled objects.

Well, yeah, that example relies on knowledge of V8 and ClearScript details that aren't very obvious 😉

First, all V8 objects are basically the same, including .NET proxies. Despite having some special wiring, .NET proxies are still JavaScript objects; that is, they have prototypes and adhere to JavaScript property access rules. In this respect V8 is very different from JScript, where host objects are fundamentally different and don't support JavaScript semantics at all.

Second, the call to `setPrototypeOf` directly from the host has the intended effect only because ClearScript caches the `dc` proxy constructed for the `AddHostObject` call. Without this caching, you'd have to call `setPrototypeOf` from script code, which would make the example code _slightly_ more complicated.

Finally, the situation here is a bit odd in that the final object and its prototype are different marshaled views of the same .NET object. This might be unusual, but it's really nothing new for JavaScript's ultra-flexible object system.

Cheers!

**Comment by KurtGokhan** [comment]
@ClearScriptLib  reopening this issue to bring this to your attention. Sorry if this question thread is getting too long.

I tried what you suggested, and I have mixed results about it. To achieve the ideal result, I wanted to automate the `setPrototypeOf` inside `OnExposedToScriptCode`. It did not work. Maybe that is expected. But there is something weird about it. See my test case:

```cs
        public class TestDictionary : TestDictionaryBase, IPropertyBag, IScriptableObject
        {
            HashSet<ScriptEngine> exposedEngines = new HashSet<ScriptEngine>();  // TODO: Weak Set
            public object prototype;
            bool setPrototypeOnExpose;

            public TestDictionary(bool setPrototypeOnExpose)
            {
                this.setPrototypeOnExpose = setPrototypeOnExpose;
            }
            public void OnExposedToScriptCode(ScriptEngine engine)
            {
                if (exposedEngines.Add(engine))
                {
                    prototype = this.ToRestrictedHostObject<TestDictionaryBase>(engine);
                    if (setPrototypeOnExpose)
                        engine.Script.Object.setPrototypeOf(this, prototype);
                }
            }
        }

        [TestMethod, TestCategory("BugFix")]
        [DataRow(true, false)] // fails
        [DataRow(false, true)]
        [DataRow(true, true)] // fails
        public void BugFix_SettingPrototypeofPropertyBag(bool setPrototypeOnExpose, bool setPrototypeAfterExpose)
        {
            var dc = new TestDictionary(setPrototypeOnExpose);
            engine.AddHostObject("dc", dc);

            if (setPrototypeAfterExpose)
                engine.Script.Object.setPrototypeOf(dc, dc.prototype);

            Assert.AreEqual("test1", engine.Evaluate("dc.TestValue")); // Fails in case 1 and 3
            engine.Execute("dc.TestValue = 123");
            Assert.AreEqual(123, dc["TestValue"]);
        }
```

Here I am trying to set the prototype in 3 different ways. 

- Case 1: During the `OnExposedToScriptCode`
- Case 2: After the `OnExposedToScriptCode`
- Case 3: Both during and after

Case 1 fails, and maybe that is normal. But I can't understand why Case 3 fails. Do you have any thoughts about this?

The reason I want to automate this is because, sometimes an object is exposed to script without my control. Also, Javascript allows the ambiguity you talked about, so it seems natural to me to do this.

![image](https://user-images.githubusercontent.com/6034931/125295598-1083b580-e32e-11eb-8de0-c12e9b1132f5.png)

**Comment by ClearScriptLib** [comment]
Hi @KurtGokhan,

Unfortunately ClearScript currently calls `OnExposedToScriptCode` before the proxy for the object is cached, so the `setPrototypeOf` call creates and modifies a redundant proxy. That second proxy then ends up in the cache, which explains the failure of case 3.

We consider this a bug. Look for a fix in the next ClearScript release.

Thanks!

**Comment by ClearScriptLib** [comment]
Fixed in [Version 7.1.5](https://github.com/microsoft/ClearScript/releases/tag/7.1.5).
